fix!: session ID stability while maintaining inactivity timeout

ttak-apphelix opened the pull request:
BREAKING CHANGE: The breaking change is that SESSION_ACTIVITY_SAVE_DELAY_SECONDS was introduced with a 15 minute default, which will change the current behavior. It is not necessarily breaking (since it actually fixes an issue), but this is to bring more attention to the new setting because the default is reasonable, but also somewhat arbitrary.

This change modifies the SessionInactivityTimeout middleware to prevent the session ID from changing on every request while still enforcing the inactivity timeout.

Key improvements:
- Store datetime values as ISO strings for proper serialization
- Implement hybrid session save approach that only allows full session saves periodically (controlled by SESSION_SAVE_FREQUENCY_SECONDS)
- Preserve session ID between requests while still tracking user activity

This resolves the issue where lms_sessionid was changing on every user interaction, which caused problems.

Private Related PR:
https://github.com/edx/edge-internal/pull/883
https://github.com/edx/edx-internal/pull/12927

Private JIRA link:
https://2u-internal.atlassian.net/browse/BOMS-66

ttak-apphelix commented:
@robrap  Can you please have a look into that

ttak-apphelix committed:
fix: session ID stability while maintaining inactivity timeout

This change modifies the SessionInactivityTimeout middleware to prevent
the session ID from changing on every request while still enforcing the
inactivity timeout.

Key improvements:
- Store datetime values as ISO strings for proper serialization
- Implement hybrid session save approach that only allows full session
  saves periodically (controlled by SESSION_SAVE_FREQUENCY_SECONDS)
- Preserve session ID between requests while still tracking user activity

This resolves the issue where lms_sessionid was changing on every user
interaction, which caused problems.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -38,23 +41,48 @@ def process_request(self, request):
 
         # Do we have this feature enabled?
         if timeout_in_seconds:
+            frequency_time_in_seconds = getattr(settings, "SESSION_SAVE_FREQUENCY_SECONDS", 60)
```
- See https://code-annotations.readthedocs.io/en/latest/contrib/how_to/documenting_django_settings.html for to annotate.
  - Annotation description should also note the new default. 
  - Consider documenting `SESSION_INACTIVITY_TIMEOUT_IN_SECONDS` as well.
- I'm going to consider this a fix, instead of a breaking change. I was debating whether we should skip the default in order to not have a breaking change, but I think it is a required improvement.

ttak-apphelix replied:
Added

robrap replied:
[proposal] 
- Change the default to 900 (15 minutes).
- The annotation could explain:
  -  The adjusted timeout would be `SESSION_INACTIVITY_TIMEOUT_IN_SECONDS` - `SESSION_SAVE_FREQUENCY_SECONDS`.

ttak-apphelix replied:
Added

robrap reviewed (commented):
@UsamaSadiq: I didn't get to the heart of this review yet. You are welcome to review if you get to it. Ultimately we'll need tests, but review the proposal first.

ttak-apphelix committed:
fix: linter issue

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -38,23 +42,48 @@ def process_request(self, request):
 
             # Get the last time user made a request to server, which is stored in session data
-            last_touch = request.session.get(LAST_TOUCH_KEYNAME)
+            last_touch_str = request.session.get(LAST_TOUCH_KEYNAME)
```
This will cause a breaking change, which I am ok with. We'll inform about this. Can you follow https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html for breaking changes, adding the `!` to PR title and commit message, and update the commit message as detailed in the OEP.

ttak-apphelix replied:
Done

robrap reviewed (commented):
Apologies. I ran out of time again. This is generally in the right direction, so please add unit tests as well. Thank you.

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix!: session ID stability while maintaining inactivity timeout

This change modifies the SessionInactivityTimeout middleware to prevent the session ID from changing on every request while still enforcing the inactivity timeout.

ttak-apphelix committed:
fix!: session ID stability while maintaining inactivity timeout

This change modifies the SessionInactivityTimeout middleware to prevent the session ID from changing on every request while still enforcing the inactivity timeout.

ttak-apphelix commented:
@robrap Added unit test files for the Session middleware

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
+        # .. setting_name: SESSION_INACTIVITY_TIMEOUT_IN_SECONDS
+        # .. setting_default: None
+        # .. setting_description:  Used for maintaining session inactivity up to n hours
+        # .. setting_warning:  Keep in sync with SESSION_COOKIE_AGE and SESSION_SAVE_FREQUENCY_SECONDS
```
```suggestion
        # .. setting_description: If set, this is used to end the session when there is no activity for N seconds.
        # .. setting_warning: Keep in sync with SESSION_COOKIE_AGE and must be larger than SESSION_SAVE_FREQUENCY_SECONDS
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
         # Do we have this feature enabled?
         if timeout_in_seconds:
+            # .. setting_name: SESSION_SAVE_FREQUENCY_SECONDS
+            # .. setting_default: 900
```
Should we make the default more like 15 minutes (in seconds)?

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
+            # .. setting_name: SESSION_SAVE_FREQUENCY_SECONDS
+            # .. setting_default: 900
+            # .. setting_description: How often to allow a full session save (in seconds).
+            #   This controls how frequently the session ID might change
```
```suggestion
            #   This controls how frequently the session ID might change.
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
+            #   A user could be inactive for almost SESSION_SAVE_FREQUENCY_SECONDS but since their session 
+            #   isn't being saved during that time, their last activity timestamp isn't being updated.
+            #   When they hit the inactivity timeout, it will be based on the last saved activity time
+            #   So the effective timeout could be as short as: SESSION_INACTIVITY_TIMEOUT_IN_SECONDS - SESSION_SAVE_FREQUENCY_SECONDS
```
```suggestion
            #   So the effective timeout could be as short as: SESSION_INACTIVITY_TIMEOUT_IN_SECONDS - SESSION_SAVE_FREQUENCY_SECONDS.
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
+            #   When they hit the inactivity timeout, it will be based on the last saved activity time
+            #   So the effective timeout could be as short as: SESSION_INACTIVITY_TIMEOUT_IN_SECONDS - SESSION_SAVE_FREQUENCY_SECONDS
+            #   This means users might be logged out earlier than expected in some edge cases.
+            # .. setting_warning:  Used in sync with SESSION_INACTIVITY_TIMEOUT_IN_SECONDS
```
```suggestion
            # .. setting_warning:  Must be smaller than SESSION_INACTIVITY_TIMEOUT_IN_SECONDS.
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
+            #   So the effective timeout could be as short as: SESSION_INACTIVITY_TIMEOUT_IN_SECONDS - SESSION_SAVE_FREQUENCY_SECONDS
+            #   This means users might be logged out earlier than expected in some edge cases.
+            # .. setting_warning:  Used in sync with SESSION_INACTIVITY_TIMEOUT_IN_SECONDS
+            frequency_time_in_seconds = getattr(settings, "SESSION_SAVE_FREQUENCY_SECONDS", 900)
```
See other PR for proposed name change.

ttak-apphelix replied:
updated the name

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -14,12 +14,16 @@
 from django.contrib import auth
 from django.utils.deprecation import MiddlewareMixin
 
 LAST_TOUCH_KEYNAME = 'SessionInactivityTimeout:last_touch'
```
Would it make sense to change the key, since the format of the values will be different?
```suggestion
LAST_TOUCH_KEYNAME = 'SessionInactivityTimeout:last_touch_str'
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
+                        return
+                except (ValueError, TypeError) as e:
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
```
Should we log a warning? This is unexpected, and a bug would cause this middleware to stop working correctly.

ttak-apphelix replied:
added

robrap replied:
Thanks. This feels like it should be `log.error` or `log.warning`. It would indicate some sort of bug that we don't want swept under the table, and we don't get debug logs in production.

ttak-apphelix replied:
updated to warning

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
+            else:
+                LOG.info("No previous activity timestamp found (first login)")
```
1. I think this would cause too mamy log messages at info level.
```suggestion
                LOG.debug("No previous activity timestamp found (first login)")
```
2. You could use `set_custom_attribute` if you want to add additional observability. That's up to you. Some possible keys include:
  - `session_inactivity.activity_seen`
  - `session_inactivity.first_login`
  - `session_inactivity.session_extended`
See https://github.com/openedx/auth-backends/pull/382 for examples.

ttak-apphelix replied:
updated

robrap reviewed (commented):
Thanks. I didn't get to unit tests, but this is a start.

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: update session inactivity middleware to use consistent naming and improve logging levels

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,71 @@ def process_request(self, request):
         # Do we have this feature enabled?
         if timeout_in_seconds:
+            # .. setting_name: SESSION_ACTIVITY_SAVE_DELAY_SECONDS
+            # .. setting_default: 15 minutes (in seconds)
```
Note: These annotations get extracted to documentation as well, and will be seen independent of the code.
```suggestion
            # .. setting_default: 900 (15 minutes in seconds)
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -14,12 +14,17 @@
+LAST_TOUCH_KEYNAME = 'SessionInactivityTimeout:last_touch_str'
+LAST_SESSION_SAVE_TIME_KEYNAME = 'SessionInactivityTimeout:last_session_save_time'
+
+LOG = logging.getLogger(__name__)
```
Nit: `LOG` is sometimes used, but lowercase `log` is consistent with most usage in this repo.

ttak-apphelix replied:
Updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,71 @@ def process_request(self, request):
+                    last_touch = datetime.fromisoformat(last_touch_str)
+
+                    # compute the delta since last time user came to the server
+                    time_since_last_activity = utc_now - last_touch
```
Try to avoid comments where your code is self-commenting (e.g. clear variables names, etc.)
```suggestion
                    last_touch = datetime.fromisoformat(last_touch_str)
                    time_since_last_activity = utc_now - last_touch
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,71 @@ def process_request(self, request):
+
+                    # did we exceed the timeout limit?
+                    if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
+                        # yes? Then log the user out
```
This is another comment where the code itself is clear enough.
```suggestion
                        # yes? Then log the user out
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,71 @@ def process_request(self, request):
+                    # compute the delta since last time user came to the server
+                    time_since_last_activity = utc_now - last_touch
+
+                    # did we exceed the timeout limit?
```
This is a useful comment. An alternative, which I am not exactly proposing, would be:
```
has_exceeded_timeout_limit = time_since_last_activity > timedelta(seconds=timeout_in_seconds):
if has_exceeded_timeout_limit:
   ...
``` 
The theory around this is that comments can get out of date and out of sync with code during refactors, but the code can't get out of sync with itself.

ttak-apphelix replied:
updated

robrap replied:
The point of this code change was to remove the comment "# did we exceed the timeout limit?", since that comment is already indicated by the variable name `has_exceeded_timeout_limit`. There are theories that it is easier for comments to get out dated, and potentially be wrong or misleading, than the code itself. So using clear variable names is an alternative way to comment your code that is less likely to get out of date.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,71 @@ def process_request(self, request):
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
+            else:
+                # Enable NR tracing for this view based on course
```
Here is a great example of a comment that is out of sync with the code. Let's just remove it, rather than trying to fix it.
```suggestion
                # Enable NR tracing for this view based on course
```

ttak-apphelix replied:
removed

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,71 @@ def process_request(self, request):
+            monitoring_utils.set_custom_attribute('session_inactivity.activity_seen', current_time)
+            if not last_save or (
+                datetime.fromisoformat(last_save) + timedelta(seconds=frequency_time_in_seconds) < datetime.utcnow()
+            ):
```
[Note to self:]
- This is where I need to pick up the review.
- The code/comments weren't immediately clear, but no changes are being requested until I can continue.

robrap replied:
Marking as resolved. I added other comments to clarify.

robrap reviewed (commented):
Submitting a review that I was unable to complete, but at least you have some of my notes. Thank you.

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: correct logging variable name and improve log levels in session inactivity middleware

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,67 @@ def process_request(self, request):
+                # Don't set modified=was_modified here! Let Django save the session
+            else:
+                # Only prevent saving in this branch
+                request.session.modified = False
```
This feels unsafe, because you don't know if changes were made outside this middleware that require the session to be saved. Instead of saving the last session save time in the session itself, you could instead use the TieredCache object (see examples throughout the codebase), which will store in shared memcached across servers.

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,67 @@ def process_request(self, request):
+            # Store activity timestamp
+            request.session[LAST_TOUCH_KEYNAME] = utc_now.isoformat()
+
+            # Periodically allow a full save (every n seconds)
```
1. This comment doesn't feel appropriate to these particular lines.
2. I'd replace comments with clear variable names like `has_exceeded_timeout_limit` above.
  - Consider `proceed_with_period_save` as one variable.
  - Maybe `has_save_delay_been_exceeded` (or something like that, could be an additional variable?

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,520 @@
+        response = self.middleware.process_request(self.request)
+        
+        assert response is None
+        assert self.request.session.modified is False
```
1. I'm not going to review unit tests in-depth until the logic gets fixed to avoid marking the session modified as False.
2. Consider using exploring `@ddt.ddt`, `@ddt.data`, `ddt.unpack` (see other tests as examples) to avoid some test redundancy.

robrap replied:
Reminder about recommendation from point 2 when you do update unit tests.

robrap replied:
@ttak-apphelix: I think you should update unit tests with your next pass of changes. Thank you.
**UPDATE:** Also, please ensure the tests are all green before requesting re-review.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,67 @@ def process_request(self, request):
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
+            else:
+                monitoring_utils.set_custom_attribute('session_inactivity.first_login', True)
```
These are missing annotations/documentation. Something like the following should be immediately above this line.
```
# .. custom_attribute_name: session_inactivity.first_login
# .. custom_attribute_description: <INSERT DESCRIPTION>.
```

ttak-apphelix replied:
Added

robrap reviewed (commented):
Thank you. Here is another batch of comments.

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: enhance session inactivity tracking with TieredCache and custom monitoring attributes

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -82,24 +82,25 @@ def process_request(self, request):
-
-            # Periodically allow a full save (every n seconds)
-            last_save = request.session.get(LAST_SESSION_SAVE_TIME_KEYNAME)
+            cached_response = TieredCache.get_cached_response(LAST_SESSION_SAVE_TIME_KEYNAME)
```
The cache key will need to include the user id so each user doesn't clobber the cached value of other users.

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -82,24 +82,25 @@ def process_request(self, request):
             ):
                 # Allow a full session save periodically
-                request.session[LAST_SESSION_SAVE_TIME_KEYNAME] = current_time
+                TieredCache.set_all_tiers(LAST_SESSION_SAVE_TIME_KEYNAME, current_time, frequency_time_in_seconds)
```
I think `frequency_time_in_seconds` is cutting it too close for the cache time. What about using `timeout_in_seconds`? Thoughts?

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +40,67 @@ def process_request(self, request):
+            # .. setting_warning:  Must be smaller than SESSION_INACTIVITY_TIMEOUT_IN_SECONDS.
+            frequency_time_in_seconds = getattr(settings, "SESSION_ACTIVITY_SAVE_DELAY_SECONDS", 900)
             # what time is it now?
             utc_now = datetime.utcnow()
```
Since you define this below, should we just rename to `current_time` everywhere for consistency, and move this outside (above) the if-block?
```suggestion
            current_time = datetime.utcnow()
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +40,67 @@ def process_request(self, request):
+
+            cached_response = TieredCache.get_cached_response(LAST_SESSION_SAVE_TIME_KEYNAME)
+            last_save = cached_response.value if cached_response.is_found else None
+            current_time = datetime.utcnow().isoformat()
```
For consistency (see other comment), this could change to something like:
```suggestion
            current_time_str = current_time.isoformat()
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +40,67 @@ def process_request(self, request):
+            # .. custom_attribute_description: Records the timestamp when user activity is detected.
+            monitoring_utils.set_custom_attribute('session_inactivity.activity_seen', current_time)
+            if not last_save or (
+                datetime.fromisoformat(last_save) + timedelta(seconds=frequency_time_in_seconds) < datetime.utcnow()
```
1. Should we switch `datetime.utcnow()` to `current_time` for consistency?
2. To make my comment https://github.com/openedx/edx-platform/pull/36896#discussion_r2167654466 from the other day more explicit, I was saying you could replace comments with clear variable names, like the following:
```
has_save_delay_been_exceeded = datetime.fromisoformat(last_save) + timedelta(seconds=frequency_time_in_seconds) < current_time
proceed_with_period_save = not last_save or has_save_delay_been_exceeded
if proceed_with_period_save:
   ...
```
Note: I switched from `datetime.utcnow()` to `current_time` again for consistency.

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +40,67 @@ def process_request(self, request):
+                monitoring_utils.set_custom_attribute('session_inactivity.first_login', True)
+                log.debug("No previous activity timestamp found (first login)")
+
+            cached_response = TieredCache.get_cached_response(LAST_SESSION_SAVE_TIME_KEYNAME)
```
[inform] TieredCache has a clear function (I forget the name) that can be used during test setup to ensure the cache has been cleared before each test.

robrap reviewed (commented):
Again, I'll wait to review unit tests once this settles a bit more. Thank you.

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: update session inactivity tracking to use user-specific cache keys and improve timestamp handling

ttak-apphelix commented:
@robrap  Changes are pushed.
I have not updated the test cases. Once implementation is verified, I'll make changes in test cases

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,520 @@
+
+    @override_settings(
+        SESSION_INACTIVITY_TIMEOUT_IN_SECONDS=300,
+        SESSION_SAVE_FREQUENCY_SECONDS=600
```
I know the unit tests have not yet been updated. This is a reminder that this name no longer exists.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -14,18 +14,28 @@
     Middleware class to keep track of activity on a given session
     """
+
+    def _get_last_session_save_time_key(self, user_id):
```
Nit: Maybe use `_get_last_session_save_time_key_cache_key` (to differentiate from session key). Not a blocker. Also, could add a short docstring for how this is used.

ttak-apphelix replied:
removed. Not required as we are removing cache

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +44,67 @@ def process_request(self, request):
+                except (ValueError, TypeError) as e:
+                    log.warning("Parsing last touch time failed: %s", e)
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
```
I don't _think_ that `pass` is needed, because the block isn't empty. I also added the comment up top (and edited it a bit due to the reordering.
```suggestion
                    # If parsing fails, log warning and then treat as if no timestamp exists
                    log.warning("Parsing last touch time failed: %s", e)
```

ttak-apphelix replied:
corrected

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +44,67 @@ def process_request(self, request):
+                # .. custom_attribute_name: session_inactivity.first_login
+                # .. custom_attribute_description: Tracks when users have no stored activity
+                #   timestamp (first access after login).
+                monitoring_utils.set_custom_attribute('session_inactivity.first_login', True)
```
One possibility is to use the following instead:
```suggestion
                monitoring_utils.set_custom_attribute('session_inactivity.last_touch_status', 'first-login`)
```
Then you could also use the following above:
```
monitoring_utils.set_custom_attribute('session_inactivity.last_touch_status', 'last-touch-exceeded`)
monitoring_utils.set_custom_attribute('session_inactivity.last_touch_status', 'last-touch-not-exceeded`)
monitoring_utils.set_custom_attribute('session_inactivity.last_touch_status', 'last-touch-error`)
```
Note: Only document with annotation before the first time you refer to this attribute key.

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -14,18 +14,28 @@
 
-LAST_TOUCH_KEYNAME = 'SessionInactivityTimeout:last_touch'
+LAST_TOUCH_KEYNAME = 'SessionInactivityTimeout:last_touch_str'
+LAST_SESSION_SAVE_TIME_KEYNAME = 'SessionInactivityTimeout:last_session_save_time:user_{}'
```
**IMPORTANT**: Please read this comment first, because it may affect future comments.

Originally, you were updating the session's `last_touch_str` and telling Django not to save the session. I noted that this was not ideal, and that if we needed to update something without updating the session, we could use a cache.

Now looking at the changes, I'm wondering if we don't need the cache, because we could simply get away with not updating anything for this case.  Basically, it doesn't seem like there is any difference between the `last_touch_str` stored in the session and the `last_session_save_time` stored in the cache. Isn't the session's `last_touch_str` the same as the time that we last saved? Does this mean we can just drop saving this redundantly to the cache?

It seems we could simply check the difference between the session's `last_touch_str` and the current time to see if there has been a long enough delay to update and save the session's `last_touch_str`, right? No need for any cache.

Apologies that I sent you down the cache front without thinking through what you were trying to update without saving. It didn't occur to me at the time that you could have simply skipped the update altogether.

ttak-apphelix replied:
You're absolutely right! Looking at the current implementation, there is indeed redundancy between last_touch_str (stored in session) and last_session_save_time. Updated that change
Yes, we can drop the cache and have updated the changes which are inline with removal of cache.

robrap reviewed (commented):
Thanks, and apologies for sending you down the wrong path (I think). At least you learned about the TieredCache? ;) 

See comments for details, and read the comment marked **important** first. I'm glad the unit tests weren't updated yet. :)

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: simplify session save time key handling and improve logging for inactivity tracking

ttak-apphelix commented:
@robrap  and @UsamaSadiq  Changes are pushed. have a look.
Note: Test cases are yet to be updated with latest changes.

ttak-apphelix committed:
fix: remove unused last session save time key and streamline session save logic

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,63 @@ def process_request(self, request):
+                # .. custom_attribute_name: session_inactivity.last_touch_status
+                # .. custom_attribute_description: Tracks the status of session activity timestamps.
+                #   Values: 'first-login' (Tracks when users have no stored activity), 'last-touch-error' (failed to parse timestamp),
+                #   'last-touch-exceeded' (Marks when sessions are extended through the periodic save), 'last-touch-not-exceeded' (within save delay).
```
Nit: This can be made a little more readable.
```suggestion
                #   Values: 'first-login' (Tracks when users have no stored activity),
                #     'last-touch-error' (failed to parse timestamp),
                #     'last-touch-exceeded' (Marks when sessions are extended through the periodic save),
                #     'last-touch-not-exceeded' (within save delay).
```

ttak-apphelix replied:
updated

robrap replied:
@ttak-apphelix: I don't think you pushed your latest changes to github, so I was unable to review.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,63 @@ def process_request(self, request):
+                request.session[LAST_TOUCH_KEYNAME] = current_time_str
+                monitoring_utils.set_custom_attribute('session_inactivity.last_touch_status', 'last-touch-exceeded')
+            else:
+                monitoring_utils.set_custom_attribute('session_inactivity.last_touch_status', 'last-touch-not-exceeded')
```
This is tracking `proceed_with_period_save`, not `has_save_delay_been_exceeded`. The value `last-touch-not-exceeded` would overwrite `first-login` above, I think.

For custom attributes, you could have:
- A single key and different string values, like we do here, but the values need to be mutually exclusive. 
- Or, you could have custom attributes with boolean values. For those, you could sometimes just set in the True case. An example would be `set_custom_attribute('session_inactivity.first_login, True)`.
- Or you could have one (or more) of each of these types.

Updating the unit tests with the custom attributes will ensure your design for observability is working.

ttak-apphelix replied:
updated with 3rd option

robrap replied:
Note: this custom attribute isn't 100% accurate, because `last_touch_status' would be 'last-touch-exceeded', even for the `not last_touch_str` case. 

If you used the following, it would provide visibility into everything and use the same names as the variables, so you wouldn't need to think about how to interpret it:
```
monitoring_utils.set_custom_attribute('session_inactivity.has_save_delay_been_exceeded', has_save_delay_been_exceeded)
monitoring_utils.set_custom_attribute('session_inactivity.proceed_with_period_save', proceed_with_period_save)
```

ttak-apphelix replied:
> Note: this custom attribute isn't 100% accurate, because `last_touch_status' would be 'last-touch-exceeded', even for the `not last_touch_str` case.
> 
> If you used the following, it would provide visibility into everything and use the same names as the variables, so you wouldn't need to think about how to interpret it:
> 
> ```
> monitoring_utils.set_custom_attribute('session_inactivity.has_save_delay_been_exceeded', has_save_delay_been_exceeded)
> monitoring_utils.set_custom_attribute('session_inactivity.proceed_with_period_save', proceed_with_period_save)
> ```

I wanted to keep the flag name same since this is in a if else case and it only denotes whether last touch exceeded or not.
Also in data dog filter it might be easier to search with single flag name to analyse and understand the behaviour of last_touch_status.
Let me know if this make sense.

robrap replied:
Thanks @ttak-apphelix. I think you are misunderstanding my point. It's minor, but when the attribute is set to `'last-touch-exceeded'`, that may not be true. In the case of first login, the attribute will also get `'last-touch-exceeded'`, but it hasn't really been exceeded. But, it will be saved.

If you only want an attribute for the if/else, then just using the following instead would be 100% accurate. It is 100% accurate because it exactly matches the condition of the if-statement you are trying to observe:
```
monitoring_utils.set_custom_attribute('session_inactivity.proceed_with_period_save', proceed_with_period_save)
```

Hopefully this is clear? If not, we'll need to discuss in person, or you can have someone else local read this and see if that helps.

ttak-apphelix replied:
Thanks @robrap for clarification. Will go ahead with 
```
monitoring_utils.set_custom_attribute('session_inactivity.proceed_with_period_save', proceed_with_period_save) 
```

robrap reviewed (commented):
It's much closer. Thanks.

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: enhance session inactivity tracking with improved logging and error handling

mraman-2U commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,67 @@ def process_request(self, request):
+
+            current_time_str = current_time.isoformat()
+
+            monitoring_utils.set_custom_attribute('session_inactivity.activity_seen', current_time_str)
```
Add comments about this custom attribute

ttak-apphelix replied:
added

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: linting fix

ttak-apphelix commented:
Internal review is completed and changes are also incorporated.
@robrap  I have updated the changes requested by you. Also, unit test cases are also updated.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +39,73 @@ def process_request(self, request):
+                    log.warning("Parsing last touch time failed: %s", e)
+                    # .. custom_attribute_name: session_inactivity.last_touch_error
+                    # .. custom_attribute_description: Boolean. True if parsing the last activity timestamp failed for this request, indicating a session data error.
+                    monitoring_utils.set_custom_attribute('session_inactivity.last_touch_error', True)
```
I didn't mean all custom attributes needed to be booleans. This was a perfectly good use of a string. :)
```suggestion
                    monitoring_utils.set_custom_attribute('session_inactivity.last_touch_error', str(e))
```

ttak-apphelix replied:
> I didn't mean all custom attributes needed to be booleans. This was a perfectly good use of a string. :)

Is this required to add exception in monitoring. As except block handles robust error handling for cases where the stored timestamp data is corrupted, malformed, or otherwise unparseable. 
If you are looking where someone is looking into Datadog and needs to understand details of the error then I can update with changes recommended by you.

robrap replied:
Since you are handling the error, DD will not get it and no one would know about it. My suggestion would provide some of the error details, rather than True.

Additionally, and even better, it is probably worth-while using [record_exception](https://github.com/search?q=org%3Aopenedx%20record_exception&type=code), to sent it to DD. Note: this only needs to be done where an exception is being handled **and** it is still unexpected and warrants looking into.

ttak-apphelix replied:
> Since you are handling the error, DD will not get it and no one would know about it. My suggestion would provide some of the error details, rather than True.
> 
> Additionally, and even better, it is probably worth-while using [record_exception](https://github.com/search?q=org%3Aopenedx%20record_exception&type=code), to sent it to DD. Note: this only needs to be done where an exception is being handled **and** it is still unexpected and warrants looking into.

Thanks, will add error details for this.

Will ignore record_exeception for this as it can be triggered for corrupted, malformed, or otherwise unparseable timestamp which does not warrant us to look into that.

Let me know your thoughts on this. If you feel otherwise I can add this.

robrap replied:
We'll use `record_exception` to know if and when a bug gets introduced, because this really shouldn't fire. You can leave the redundant custom attribute or not for the error as you see fit.

ttak-apphelix replied:
added record_exception()

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,255 @@
+        )
+        mock_monitoring.set_custom_attribute.assert_any_call(
+            "session_inactivity.last_touch_status", "last-touch-exceeded"
+        )
```
Nit: See example that is less repetitive and simpler to read: https://github.com/openedx/edx-platform/blob/0913b51170f75d14d739f5aab6dc2c9a61f324ea/openedx/core/lib/x_forwarded_for/tests/test_middleware.py#L83-L88

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,255 @@
+        (300, 6, True),
+        (300, 5, False),
+        (600, 9, False),
+        (600, 11, True),
```
Minor points:
1. Why is minutes_elapsed not ordered? Just makes it more challenging to read.
2. Why not use seconds_elapsed, so you don't need to convert in your head?
3. Consider combining the following tests and using, so you can see the interplay of the two:
```
# (timeout_seconds, save_delay_seconds, minutes_elapsed, should_logout)
```
4. Aren't there some custom attributes to be checked as well?

ttak-apphelix replied:
1. Updated the order
2. changes to seconds
3. merged to test cases
4. Added custom attributes for unit test.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,255 @@
+        mock_datetime.utcnow.return_value = mock_now
+
+        with self.assertRaises((ValueError, TypeError)):
+            self.middleware.process_request(self.request)
```
Maybe I'm missing something, but I thought you had coded this to log a warning and a custom attribute and move on, rather than raising an error? What is the graceful handling?

ttak-apphelix replied:
Correct. Removed test case as it was not relevant now.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,255 @@
+        # Empty string is falsy, so it should be treated as first login
+        assert self.request.session[LAST_TOUCH_KEYNAME] == mock_now.isoformat()
+        mock_monitoring.set_custom_attribute.assert_any_call(
+            "session_inactivity.first_login", True
```
How does this differ from `test_process_request_first_visit_sets_timestamp`?

ttak-apphelix replied:
Both were different use case which can be merged in single use case. Updated test_process_request_first_visit_sets_timestamp with that.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,255 @@
+
+    @override_settings(SESSION_INACTIVITY_TIMEOUT_IN_SECONDS=300)
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.datetime")
+    def test_process_request_uses_default_save_frequency(self, mock_datetime):
```
I'm on the fence about testing the defaults, but if we do, I'm not clear on how this tests it. Wouldn't we need tests on either side of the value using ddt or something?

ttak-apphelix replied:
Updated test cases with ddt with default and boundary value.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,255 @@
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.datetime")
+    def test_process_request_empty_session(self, mock_datetime):
+        # Clear the session but keep it as a proper session object
+        self.request.session.flush()
```
Curious how this case is different from the newly created session. Is there code that isn't tested without this?

ttak-apphelix replied:
Both are similar Hence removing this test case

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,255 @@
+        assert self.request.session[LAST_TOUCH_KEYNAME] == mock_now.isoformat()
+
+    @ddt.data(
+        # Test boundary conditions more precisely
```
1. I like it when you add a comment noting what each variable is.
2. Why is this a different test from the one above? The point of ddt is to add the cases you wish to test.
- Note: You can add comments above different sets of cases as well if that helps.

ttak-apphelix replied:
Missed to update the comment for this. Updated.

robrap reviewed (commented):
Apologies. Ran out of time for reviewing, but this should get you started.

ttak-apphelix commented:
https://github.com/openedx/edx-platform/pull/36896#discussion_r2237837259
@robrap Not sure why quote reply for not showing for above thread. Starting conversation here :

I wanted to keep the flag name same since this is in a if else case and it only denotes whether last touch exceeded or not.
Also in data dog filter it might be easier to search with single flag name to analyze and understand the behavior of last_touch_status.
Let me know if this make sense.

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: enhance session inactivity monitoring with improved error handling

ttak-apphelix committed:
fix: fix linter and import

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +40,75 @@ def process_request(self, request):
+
+            # .. custom_attribute_name: session_inactivity.activity_seen
+            # .. custom_attribute_description: The current timestamp when user activity was detected for this request.
+            monitoring_utils.set_custom_attribute('session_inactivity.activity_seen', current_time_str)
```
I'm not really seeing how this will be used during rollout, so I think it could be removed. If you feel strongly about it, you can leave it, but could you rename to `session_inactivity.activity_time_str`? The name `activity_seen` seems more like a boolean, and doesn't match the name of the variable at all, so something seems off.

ttak-apphelix replied:
I have removed `session_inactivity.activity_seen` monitoring.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +40,75 @@ def process_request(self, request):
+                    last_touch = datetime.fromisoformat(last_touch_str)
+                    time_since_last_activity = current_time - last_touch
+
+                    has_exceeded_timeout_limit = time_since_last_activity > timedelta(seconds=timeout_in_seconds)
```
Immediately after this line, and before the conditional, I think we'd want a custom attribute for `session_inactivity.has_exceeded_timeout_limit`. This will allow us to more easily see who is actually getting logged out due to the middleware.

ttak-apphelix replied:
add new monitoring `session_inactivity.has_exceeded_timeout_limit` and test cases are also updated accordingly

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,209 @@
+            response = self.middleware.process_request(self.request)  # lint-amnesty, pylint: disable=assignment-from-none
+
+        assert response is None
+        assert LAST_TOUCH_KEYNAME not in self.request.session
```
I don't think this test is needed. I think `test_process_request_unauthenticated_user_does_nothing` already tests `SESSION_INACTIVITY_TIMEOUT_IN_SECONDS` unset. I think `SESSION_INACTIVITY_TIMEOUT_IN_SECONDS=0` as a way to disable this is a strange exception, and should just be ignored rather than tested.
Thoughts?

ttak-apphelix replied:
I agree that test_process_request_timeout_disabled_does_nothing should be removed.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,209 @@
+
+    @ddt.data(
+        None,  # No timestamp key in session
+        "",    # Empty string timestamp
```
Consistency here will make this more clear.
```suggestion
        "",    # Empty string timestamp in session
```

ttak-apphelix replied:
updated

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,209 @@
+        with self.assertLogs(
+            "openedx.core.djangoapps.session_inactivity_timeout.middleware",
+            level="DEBUG",
+        ) as log:
```
Personal nit: (a.k.a. you do not need to update this) I find this context manager obscures the test and the `process_request` line that is actually being tested. I'd just patch mock the logging same as everything else.

ttak-apphelix replied:
updated with mock

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,209 @@
+        (3600, 900, 600, False),  # 10 min < 15 min save delay, no save
+        (3600, 900, 1200, False),  # 20 min > 15 min save delay, save occurs
+        (3600, 600, 480, False),  # 8 min < 10 min save delay, no save
+        (3600, 600, 720, False),  # 12 min > 10 min save delay, save occurs
```
Rather than have `should_logout` hard coded, and `should_save` calculated, I think the save part should move to its own test with `should_save` hard coded. This other test wouldn't need `should_logout`, since they were all False anyway.

ttak-apphelix replied:
updated and separated the test to `test_process_request_timeout_behavior` and `test_process_request_save_behavior`

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,209 @@
+
+            if should_logout:
+                assert LAST_TOUCH_KEYNAME not in self.request.session
+                mock_auth.logout.assert_called_once_with(self.request)
```
Does auth need to be mocked? Could you assert that the session has a user at top, and that it becomes anonymous?

ttak-apphelix replied:
updated with user at top and then asserting that

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,209 @@
+                    # Session should not be saved, timestamp remains the same
+                    assert self.request.session[LAST_TOUCH_KEYNAME] == last_touch.isoformat()
+                    mock_monitoring.set_custom_attribute.assert_any_call(
+                        "session_inactivity.proceed_with_period_save", False
```
Not sure if `activity_seen` is going away, but otherwise, is it not also called here?

ttak-apphelix replied:
not required now, as this is been removed

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,209 @@
+        # (seconds_elapsed, should_save) - testing around default 900 second save delay
+        (800, False),   # 13.3 min < 15 min default, no save
+        (900, False),   # exactly 15 min default, no save (not exceeded)
+        (1000, True),   # 16.7 min > 15 min default, save occurs
```
1. The only difference between testing boundary conditions and not, is the inputs. I think the boundary conditions are more important (e.g. 899,900,901), but if you want both, just have 800,899,900,901,1000 in one test. Note: For testing default setting, you could use abbreviated 899,900,901, and for testing non-default you could use the full list.
2. The code comments in minutes I find needlessly confusing. If you had `# 899 sec < 900 sec default, no save`, it really doesn't matter what this is in minutes, it is easier to match against the 899 and 900 in the test and early comment. This applies elsewhere as well.

ttak-apphelix replied:
1. Updated
2. Changed min to sec in all the comment section to avoid confusion.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/middleware.py:
```diff
@@ -34,27 +40,75 @@ def process_request(self, request):
+                        return
+                except (ValueError, TypeError) as e:
+                    # If parsing fails, log warning and then treat as if no timestamp exists
+                    log.warning("Parsing last touch time failed: %s", e)
```
Is it difficult to have a unit test with an illegal timestamp in the session?

ttak-apphelix replied:
add test case for that

robrap reviewed (commented):
I know there are a lot of comments, but this is quite close, and the tests are so much simpler to follow. Thank you!

ttak-apphelix committed:
fix: incorporated review comment

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix: removed unwanted comments

ttak-apphelix committed:
fix: updated mock for log

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -68,47 +60,61 @@ def test_process_request_first_visit_sets_timestamp(
         assert response is None
         assert self.request.session[LAST_TOUCH_KEYNAME] == mock_now.isoformat()
 
+        # Verify debug log is called for first login scenarios
```
Nit: the line speaks for itself, and comment is not necessary.

ttak-apphelix replied:
comment is removed

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -121,31 +127,75 @@ def test_process_request_timeout_and_save_behavior(
+        (600, 599, False),   # 599 sec < 600 sec save delay, no save
+        (600, 600, False),   # 600 sec = 600 sec save delay, no save (not exceeded)
+        (600, 601, True),    # 601 sec > 600 sec save delay, save occurs
+        (600, 720, True),    # 720 sec > 600 sec save delay, save occurs
```
[food for thought] The number of tests means more for a human to understand, and also takes more processing time to run. I would question whether 600 is really any different than 899 test, and if 899 is good enough? Same with other non-boundaries.

ttak-apphelix replied:
updated non-boundaries cases

robrap replied:
[inform] I added the same change here as well: 
- https://github.com/openedx/edx-platform/pull/36896#discussion_r2289657879

robrap reviewed (approved):
Looks great. Thanks.
- I did leave some minor comments for you to consider. They are optional.
- There is a minor quality failure to be cleaned up.
- Can you confirm that this was tested locally (devstack) or on a sandbox, in addition to just unit tests? Thank you.

**[UPDATED]:** Additionally, we should add to the PR description and final commit comment something like the following:
```
BREAKING CHANGE: The breaking change is that SESSION_ACTIVITY_SAVE_DELAY_SECONDS was
introduced with a 15 minute default, which will change the current behavior. It is not necessarily breaking
(since it actually fixes an issue), but this is to bring more attention to the new setting because the default
is reasonable, but also somewhat arbitrary.
```

ttak-apphelix committed:
Merge branch 'openedx:master' into ttak/session_id_fix

ttak-apphelix committed:
fix!: session ID stability while maintaining inactivity timeout

BREAKING CHANGE: The breaking change is that SESSION_ACTIVITY_SAVE_DELAY_SECONDS was introduced with a 15 minute default, which will change the current behavior. It is not necessarily breaking (since it actually fixes an issue), but this is to bring more attention to the new setting because the default is reasonable, but also somewhat arbitrary.

ttak-apphelix commented:
> Looks great. Thanks.
> 
> * I did leave some minor comments for you to consider. They are optional.
> * There is a minor quality failure to be cleaned up.
> * Can you confirm that this was tested locally (devstack) or on a sandbox, in addition to just unit tests? Thank you.
> 
> **[UPDATED]:** Additionally, we should add to the PR description and final commit comment something like the following:
> 
> ```
> BREAKING CHANGE: The breaking change is that SESSION_ACTIVITY_SAVE_DELAY_SECONDS was
> introduced with a 15 minute default, which will change the current behavior. It is not necessarily breaking
> (since it actually fixes an issue), but this is to bring more attention to the new setting because the default
> is reasonable, but also somewhat arbitrary.
> ```

1. All comment are addressed.
2. Quality failure is fixed.
3. It's been tested locally. 

4. PR description and commit is been updated with suggested description.

robrap commented on openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py:
```diff
@@ -0,0 +1,219 @@
+        (300, 299, False),   # 299 sec < 300 sec timeout, no logout
+        (300, 300, False),   # 300 sec = 300 sec timeout, no logout (not exceeded)
+        (300, 301, True),    # 301 sec > 300 sec timeout, logout occurs
+        (300, 360, True),    # 360 sec > 300 sec timeout, logout occurs
```
@ttak-apphelix: I would have thought you'd apply the same removal of non-boundary tests to all tests. I'm going to commit this change now.
```suggestion
        (300, 299, False),   # 299 sec < 300 sec timeout, no logout
        (300, 300, False),   # 300 sec = 300 sec timeout, no logout (not exceeded)
        (300, 301, True),    # 301 sec > 300 sec timeout, logout occurs
```

robrap committed:
fixup! Update openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py

robrap reviewed (approved):
Yay!

robrap committed:
Merge branch 'master' into ttak/session_id_fix

robrap commented:
@ttak-apphelix: I can merge tomorrow my time once tests have passed again (or @UsamaSadiq can merge your time). Then you can enable and test on Stage. We'll need to discuss the larger rollout plan before it can move to other environments. Thank you.
